### ISSUE TITLE
mat-error is not making aria-live announcements

### ISSUE THREAD

**Issue by jelbourn** [angular/components#13952]
`MatError` has `role="alert"` applied, but this won't take effect when it's appended to the DOM. We need to either:
* Change `<mat-error>` to always be in the DOM (`display: none`) and show it when a form-field goes into error-state, or
* Use the `LiveAnnouncer` service to announce the contents (as an alert) when the control goes into an error state.

**Comment by crisbeto** [comment]
The first solution is tricky, because we don't have control over when the errors are displayed. It's up to the consumer to add the `mat-error` and toggle it using `ngIf` or `display`.

**Comment by jelbourn** [comment]
This might actually not be necessary- I got some more information about how `role="alert"` actually is supposed to behave differently from `aria-live`. I need to do some more follow-up to see if we need to make a change.

**Comment by jelbourn** [comment]
Related to #14281

**Comment by mfairchild365** [comment]
@jelbourn posting this here because it is related.  Please let me know if you want me to create a separate issue for this.

`aria-describedby` has poor support when referencing an element with `role="alert"`. Most screen reader / browser combinations will not announce the error message at all when the input receives focus. This might make it difficult for screen reader users to discover error messages as they navigate back through the form.

It's not clear if this is by design in the ARIA spec or if this due to a browser bug. See https://bugzilla.mozilla.org/show_bug.cgi?id=1505974 and https://a11ysupport.io/tests/tech__aria__aria_describedby_with_role_alert for more info

Suggest using `aria-live="assertive"` and `aria-atomic="true"` instead of `role="alert"`. This will accomplish the same thing and fix the issue mentioned above.

**Comment by ramjiseetharaman** [comment]
@mfairchild365 - I feel that since we are associating the input field with the error message using the aria-describedby attribute, we don't need to live region on the inline error message. I have had an opportunity working with users to do Usability testing on forms. One thing we inferred from that is, 70% of the users felt that the announcement of the error messages as live regions were jarring/annoying. The reasoning behind that is that, assistive technology users reported that they navigate through all the input fields first to get the feel of what is expected out of the form before they start filling out. 

With that being said, if you really want to communicate the error messages as live regions, it is recommended by W3C to use `aria-live="polite"` rather than role="alert" which implicitly corresponds to aria-live="assertive" and aria-atomic="true". Please find the link as reference.
[](https://www.w3.org/WAI/tutorials/forms/notifications/)

> Note: The displayed message in this example is coded using a <span> element that has an aria-live attribute with the value polite. The contents of this so called “live region” are conveyed to screen readers and other assistive technology. The value “polite” de-emphasizes the importance of the message and does not cause screen readers to interrupt their current tasks to read aloud this message. Thus the message is only read once when the user stops typing rather than on every keystroke that the user makes.

PS: Unfortunately, since the Usability study is a proprietary property of my company, I can't share any links here.

**Comment by ramjiseetharaman** [comment]
Got an update with reference to the Usability study I was talking about. Here is the link to the study I am talking about,
https://simplyaccessible.com/article/accessible-forms-angular/

**Comment by zarend** [comment]
closing because I believe this has been fixed by #21870
